Correct design.rst dictionaries section

viagluck opened the pull request:
The section I propose to change implies that constant time means that all keys have to have different hash values--i.e. no collisions. This is not the case.

<!--
Thanks for your contribution!
Please read this comment in its entirety. It's quite important.

# Pull Request title

It should be in the following format:

```
bpo-NNNN: Summary of the changes made
```

Where: bpo-NNNN refers to the issue number in the https://bugs.python.org.

Most PRs will require an issue number. Trivial changes, like fixing a typo, do not need an issue.

# Backport Pull Request title

If this is a backport PR (PR made against branches other than `master`),
please ensure that the PR title is in the following format:

```
[X.Y] <title from the original PR> (GH-NNNN)
```

Where: [X.Y] is the branch name, e.g. [3.6].

GH-NNNN refers to the PR number from `master`.

-->

viagluck committed:
Update design.rst

The section I propose to change implies that constant time means that all keys have to have different hash values--i.e. no collisions. This is not the case.

rhettinger commented on Doc/faq/design.rst:
```diff
@@ -499,8 +499,7 @@ using the :func:`hash` built-in function.  The hash code varies widely depending
-internal array where the value will be stored.  Assuming that you're storing
-keys that all have different hash values, this means that dictionaries take
+internal array where the value will be stored.  Hash tables, and therefore dictionaries, take
 constant time -- O(1), in Big-O notation -- to retrieve a key.
```
Sooner or later, [someone will complain](https://stackoverflow.com/questions/9214353/hash-table-runtime-complexity-insert-search-and-delete) that the new docs are incorrect because dictionaries have O(n) worst case performance., so some sort of qualifier is still necessary.  We get O(1) performance in the typical case when the hash values are evenly distributed; however, it is easy to construct degenerate cases with a catastrophic pile-ups at a single value.   

Consider using the [wikipedia wording](https://en.wikipedia.org/wiki/Hash_table) where we mention both the average case and worst case.

viagluck committed:
Update design.rst

Modified the section "How are Dictionaries implemented in CPython?" to mention a dictionary's average-case and worst-case time complexity.

csabella reviewed (changes requested):
On the first line in this section ``CPython's dictionaries are implemented as resizable hash tables.``, please add a link to `hash tables` to the wikipedia page that Raymond referenced.  Thanks!

viagluck committed:
Update design.rst

csabella reviewed (approved):
LGTM.   Thanks!
